File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/app/AppWithCircularDependency.java
package ru.tinkoff.kora.kora.app.annotation.processor.app;

import ru.tinkoff.kora.common.KoraApp;
import ru.tinkoff.kora.common.annotation.Root;

@KoraApp
public interface AppWithCircularDependency {
    @Root
    default Class1 class1(Class2 value) {
        return new Class1();
    }

    @Root
    default Class2 class2(Class3 value) {
        return new Class2();
    }

    @Root
    default Class3 class3(Class1 value) {
        return new Class3();
    }


    record Class1() {}

    record Class2() {}

    record Class3() {}

}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/app/AppWithCircularDependency.kt
package ru.tinkoff.kora.kora.app.ksp.app

import ru.tinkoff.kora.common.KoraApp
import ru.tinkoff.kora.common.annotation.Root

@KoraApp
interface AppWithCircularDependency {
    @Root
    fun class1(class2: Class2) = Class1()
    fun class2(class3: Class3): Class2
    fun class3(class1: Class1): Class3

    class Class1

    class Class2

    class Class3
}
